### PR TITLE
Docs: Add repository tutorial based on metadata API

### LINKED ISSUE #1673: Docs: Add repository tutorial based on metadata API
**Description of issue or feature request**:

For the lack of dedicated repository tooling - *`tuf/repository_tool.py` will be deprecated; alternatives are still being designed* - I suggest to create a simple document, that demonstrates how to create and maintain TUF repos, using only the metadata API.

These scenarios are described in the [advanced tutorial](https://github.com/theupdateframework/python-tuf/blob/develop/docs/TUTORIAL.md) and in [this usage scenario draft document](https://docs.google.com/document/d/1qBNR-_HuS312H9YjFdIbll8MW8Abe9ORYzx43z-bpEk). Plus the [repo editor](https://github.com/vmware-labs/repository-editor-for-tuf) demonstrates how to use the metadata API for this purpose (its generic UI aspect may be ignored).

**Current behavior**:
- No example code that shows how to create and maintain TUF repositories, using the metadata API exclusively.

**Expected behavior**:
- Example code that shows how to create and maintain TUF repositories, using the metadata API exclusively.

### PR DESCRIPTION
**Pull request by lukpueh**
*[EDIT 11/23/2021: remove draft status and update PR description]*
*[EDIT 11/25/2021: remove reviewer question about test, the PR includes one]*

Fixes #1673

**Description**

Add an excessively commented python script to demonstrate repository creation/operation using only the low-level metadata API.

**Context**
As 'repository_tool' and 'repository_lib' are being deprecated, repository metadata must to be created and maintained *manually* using the low-level Metadata API. The example code in these files shall serve as temporary replacement until a new repository tool is available.

**Contents**
 - creation of top-level metadata
 - target file handling
 - consistent snapshots
 - key management
 - top-level delegation and signing thresholds
 - target delegation
 - in-band and out-of-band metadata signing
 - writing and reading metadata files
 - root key rotation

**Questions for reviewers**
 * Is this readable enough? Should this rather be a markdown document?
 * Is it too broad? Should we move some usage scenarios to different files?
   E.g. hashed bin delegation will be in a separate file (see follow-up PR)
 * How can we demonstrate validity? With a TrustedMetadata example?
 * How can we demonstrate usage? With a client example follow-up?

**Please verify and check that the pull request fulfills the following requirements:**
- [X] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [X] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py]: Maybe we can somehow point to the Metadata class here?

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py]: I have no good suggestion for improvement but these paths seem confusing. I admit I had to read it twice even though I was supposed to know what is going on :grin: Maybe an example path and/vs URL could help.

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py]: Doesn't the client always try to download timestamp?

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py:397]: I think the part "signing with a threshold of old keys AND a threshold of the new keys" needs an emphasis.

**Review by sechkova — COMMENTED** [review]
Thanks for the refreshment exercise :) I left some comments on the text itself.
I don't think the example is too broad, having everything at one place seems like a good idea. Maybe in time, more detailed examples can be added for some trickier cases (complex delegations, key rotations ...).

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py]: I can see that how it confusing. I'll try to better explain/exemplify those two paths.

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py:397]: ➕  I absolutely agree.

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py]: Ha! So much for "fully internalized spec". Thanks for catching this. I'll reword.

**Comment by lukpueh** [comment]
Thanks for your review, @sechkova! I just finalised the example texts and squashed all commits into one. Your comments should be addressed in another commit on top of that for re-review convenience, i.e. 2fc5945.

I'd appreciate another look. Also see the updated PR description, especially the "Questions for reviewers".

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py]: nit:
```suggestion
metadata must be created and maintained *manually* using the low-level
```

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py:115]: I like it!

**Comment by lukpueh** [comment]
I just added very basic testing that runs the example script using `exec` and checks if the metadata files were created as expected. It currently only checks for file existence, but it wouldn't be hard to add more thorough inspection of the created files.

Note that the test class is intended to be re-used for a similar example script that I have in the pipe.

File: examples/repo_example/basic_repo.py
Comment (sechkova) [inline examples/repo_example/basic_repo.py]:
```suggestion
        meta={"targets.json": MetaFile(version=1)},
```
It is a nit but since it is an example ...

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Typo:
```suggestion
metadata must be created and maintained *manually* using the low-level
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: WDYT to:
```suggestion
Metadata API. The example code in this file demonstrates how to
implement similar functionality to that of the legacy 'repository_tool'
and 'repository_lib' until a new repository implementation is available.
```

Indicates it's an example (and that a user should think before copy/paste verbatim) and doesn't promise to provide a new repository *tool*.

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Is bumped obvious enough a term? If not, perhaps:
```suggestion
# The 'version', which is incremented on each metadata change, is used to
# reference metadata from within other metadata, and thus allows for repository
# consistency in addition to protecting against rollback attacks.
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: consistency:
```suggestion
# The 'expiry' date protects against freeze attacks and allows for implicit key
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Nit: use a similar phrase?
```suggestion
# keys (root, top-level targets) may have longer expiration intervals.
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Nit: might be cleaner to resolve() earlier, i.e. ln 116. It looks a bit funny doing operations on `local_path` to create `target_path` and only later resolving the full path. I understand why this works, but had to read twice to make sure it did.

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Observation: the comments earlier suggest that snapshot is usually a shorter expiration interval than targets, but the inverse is true for this code.

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]:
```suggestion
# signature thresholds is to avoid having private keys all in one place.
```

**Review by sechkova — APPROVED** [review]
Would be nice if more people provide feedback but you have my approval :grin: 
I also filled in the questionnaire.

>     Is this readable enough? Should this rather be a markdown document?
No strong opinion here, I'm ok with this one
>     Is it too broad? Should we move some usage scenarios to different files?
>     E.g. hashed bin delegation will be in a separate file (see follow-up PR)
Nope, I like it,
>     How can we demonstrate validity? With a TrustedMetadata example?
Seems like the options are either a full Updater with a server or using  TrustedMetadata by providing it with the metadata files as bytes. The second option sounds more interesting to me :) 
>     How can we demonstrate usage? With a client example follow-up?
I have nothing better to propose.

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]:
```suggestion
# thereof) to other targets roles. This allows creation of a granular trust
# hierarchy, and further reduces the impact of a single role compromise.
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: typo:
```suggestion
# https://theupdateframework.github.io/specification/latest/#delegations
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Nit: aids readability IMHO
```suggestion
        meta={"targets.json": MetaFile(version=1)},
```

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [inline examples/repo_example/basic_repo.py]: Nit: same here
```suggestion
        snapshot_meta=MetaFile(version=1),
```

File: tests/test_examples.py
Comment (joshuagl) [inline tests/test_examples.py]:
```suggestion
    the 'repo_example' directory.
```

File: tests/test_examples.py
Comment (joshuagl) [inline tests/test_examples.py]:
```suggestion
        """Locate and cache repo_example dir."""
```

File: tests/test_examples.py
Comment (joshuagl) [inline tests/test_examples.py:27]: Should we use `os.path.join()` or do we not want to run this on Windows?

File: tests/test_examples.py
Comment (joshuagl) [inline tests/test_examples.py]:
```suggestion
        """Run script in 'repo_example' dir and assert that it created the
```

**Review by joshuagl — APPROVED** [review]
Thanks for this @lukpueh, beautiful work – very clear example and a testament to the Metadata API. 🎉 

> Is this readable enough? Should this rather be a markdown document?

To me this is very readable, appreciate the detail and care here.

> Is it too broad? Should we move some usage scenarios to different files?

LGTM as is. Covers the common cases, I think.

> How can we demonstrate validity? With a TrustedMetadata example?

Implicit in the below? And future work once the proposed repository API validation piece is implemented? i.e. I think this is a solid submission as-is.

> How can we demonstrate usage? With a client example follow-up?

Could be a simple follow-on that runs a http server over the data generated here and uses ngclient to fetch the python file?

File: examples/repo_example/basic_repo.py
Comment (jku) [inline examples/repo_example/basic_repo.py]:
```suggestion
# Remove target file info from top-level targets (delegatee is now responsible)
```

File: tests/test_examples.py
Comment (jku) [inline tests/test_examples.py:42]: martin is trying to enable mypy (along with the linters) in new test files so argument type annotations would be nice -- but handling these in a follow-up issue is fine IMO.

**Review by jku — APPROVED** [review]
LGTM, I think this is fine as the stop gap: it's demonstrating the basic elements that a repository needs to handle, without pretending to actually be a repository implementation.

My reaction to the "is it too broad" question is that we shouldn't put a lot of effort into expanding or splitting it (although the hashed bin example should be useful), if we can push the repository library and its example code (or a repository tool) forward instead...

The repository validation question is more interesting... but even there maybe the answer isn't to expand this tutorial code . 

On the usage question: this could be a follow-up issue that is solved when initial versions of both this and the client example are present. Possible solution:
* document how to serve the generated directory over http
* add a "bootstrap from this root.json" option to the client so it can be used with the generated repository

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py]: The field name is 'expires'. Should we write *"The 'expires' date... "*? Does consistency beat grammar? :D 
Regardless, I'll remove the stray comma.

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py]: Agreed. Will update!

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [reply examples/repo_example/basic_repo.py]: 🤦 of course.

I think it is better to bind the comments to the variables, maybe:
"The date the metadata 'expires'..." ?

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py]: Very good observation! My (admittedly not obvious) thinking here was that the top-level targets in the example is used *like* delegated targets, i.e. to list target files, and thus might change often, and thus might expire more quickly than a top-level target file that only delegates to other targets.

Do you think it is worth including this reasoning in the example, or does it go beyond its scope? I could try to update the comments and maybe also change the expiration interval below, when I make top-level targets a pure delegator.

File: tests/test_examples.py
Comment (lukpueh) [reply tests/test_examples.py:27]: This actually works on Windows too, because `pathlib` overloads the `/` operator to make it act like `join`.

https://docs.python.org/3/library/pathlib.html#operators

File: tests/test_examples.py
Comment (lukpueh) [reply tests/test_examples.py:42]: If you really don't mind, I'll leave this to a follow-up.

File: examples/repo_example/basic_repo.py
Comment (lukpueh) [reply examples/repo_example/basic_repo.py]: @joshuagl, do you think 83764c6 is enough?

**Comment by lukpueh** [comment]
Thanks for the thorough reviews, @sechkova, @joshuagl and @jku! I addressed your comments in  4327214 and  83764c6 (the latter in response to https://github.com/theupdateframework/python-tuf/pull/1685#discussion_r756967300).

May I ask for another quick round of LGTMs? (NOTE: the commits are labeled WIP, because I intend to squash them before merging.)

**Review by jku — APPROVED** [review]
:+1:

File: tests/test_examples.py
Comment (joshuagl) [reply tests/test_examples.py:27]: Oh, wow. TIL. I cannot decide whether I like this magic or not 😆

File: examples/repo_example/basic_repo.py
Comment (joshuagl) [reply examples/repo_example/basic_repo.py]: Seems more than sufficient, thank you!

**Review by joshuagl — APPROVED** [review]
LGTM, thanks!

**Review by MVrachev — COMMENTED** [review]
I just read it and liked the comments and comparing that to `tuf/repository_tool.py` is such an improvement and simplification.
Great work!

**Comment by lukpueh** [comment]
Thanks, everyone!

**Pull request merged**
